sdk/keyvault/keyvault-keys/src/cryptographyClient.ts

```diff
@@ -0,0 +1,524 @@
+import { GetKeyOptions, RequestOptions } from "./keysModels";
+import { JsonWebKey, JsonWebKeyEncryptionAlgorithm, JsonWebKeySignatureAlgorithm } from "./core/models";
+import {
+  ServiceClientCredentials, TokenCredential, isNode, RequestPolicyFactory,
+  isTokenCredential,
+  deserializationPolicy,
+  signingPolicy,
+  exponentialRetryPolicy,
+  redirectPolicy,
+  systemErrorRetryPolicy,
+  generateClientRequestIdPolicy,
+  proxyPolicy,
+  throttlingRetryPolicy,
+  getDefaultProxySettings,
+  userAgentPolicy,
+  getDefaultUserAgentValue,
+} from "@azure/core-http";
+import { parseKeyvaultIdentifier } from "./core/utils";
+import { TelemetryOptions } from "./core";
+import { RetryConstants, SDK_VERSION } from "./core/utils/constants";
+import {
+  NewPipelineOptions,
+  isNewPipelineOptions,
+  Pipeline,
+} from "./core/keyVaultBase";
+import { KeyVaultClient } from "./core/keyVaultClient";
+import { challengeBasedAuthenticationPolicy } from "./core/challengeBasedAuthenticationPolicy";
+import * as crypto from "crypto";
+import * as constants from "constants";
+
+let keyto: any;
+if (isNode) {
+  keyto = require("@trust/keyto");
+}
+
+/**
+ * The client to interact with the KeyVault cryptography functionality
+ */
+export class CryptographyClient {
+
+  /**
+   * Retrieves the complete key from the key vault
+   * @param options Options for retrieving key
+   */
+  public async getKey(options?: GetKeyOptions): Promise<JsonWebKey> {
+    if (typeof this.key === "string") {
+      if (!this.name || this.name === "") {
+        throw new Error("getKey requires a key with a name");
+      }
+      const key = await this.client.getKey(
+        this.vaultBaseUrl,
+        this.name,
+        options && options.version ? options.version : (this.version ? this.version : ""),
+        options
+      );
+      return key.key!;
+    } else {
+      return this.key;
+    }
+  }
+
+  /**
+   * Encrypts the given plaintext with the specified cryptography algorithm
+   * @param plaintext The text to encrypt
+   * @param algorithm The algorithm to use
+   * @param options Additional options
+   */
+  public async encrypt(
+    plaintext: Uint8Array,
+    algorithm: JsonWebKeyEncryptionAlgorithm,
+    options?: EncryptOptions
+  ): Promise<Uint8Array> {
+    if (isNode) {
+      await this.fetchFullKeyIfPossible();
+
+      if (typeof this.key !== "string") {
+        switch (algorithm) {
+          case "RSA1_5": {
+            let keyPEM = keyto.from(this.key, "jwk").toString('pem', 'public_pkcs1');
+
```

Comment:
You should check the key type is appropriate for the key type.  And that it contains the key material needed to preform the encryption.  This applies to all methods here.

Reply:
All local methods or all methods?

Reply:
Yes only if you have the key locally, also if you have the key locally you should also check that the key operation is allowed in the key_ops list.

---
In reply to: [309416031](https://github.com/Azure/azure-sdk-for-js/pull/4434#discussion_r309416031) [](ancestors = 309416031)